### PR TITLE
Implement multi insert test case for pkg agent handler

### PR DESCRIPTION
**Pull request by kevindiu**
Signed-off-by: kevindiu <kevindiujp@gmail.com>

### Description:

This PR implements MultiInsert test for pkg agent handler.

Part 1: https://github.com/vdaas/vald/pull/1630 merged
Part 2:https://github.com/vdaas/vald/pull/1645 merged
Part 3: https://github.com/vdaas/vald/pull/1646 merged
Part 4: https://github.com/vdaas/vald/pull/1651 merged

### Environment:

- Go Version: 1.18
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 1]
```diff
@@ -3385,6 +3385,69 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				- case 4.4: MultiInsert 100 vector with 50 request with empty UUID fail (vector type is float32)
+				- case 4.5: MultiInsert 100 vector with all request with empty UUID fail (vector type is uint8)
+				- case 4.6: MultiInsert 100 vector with all request with empty UUID fail (vector type is float32)
+				- case 5.1: MultiInsert 100 vector with 1 vector with maximum dimension fail (vector type is uint8)
+				- case 5.2: MultiInsert 100 vector with 1 vector with maximum dimension fail (vector type is float32)
+				- case 5.3: MultiInsert 100 vector with 50 vector with maximum dimension fail (vector type is uint8)
+				- case 5.4: MultiInsert 100 vector with 50 vector with maximum dimension fail (vector type is float32)
+				- case 5.5: MultiInsert 100 vector with all vector with maximum dimension fail (vector type is uint8)
+				- case 5.6: MultiInsert 100 vector with all vector with maximum dimension fail (vector type is float32)
+
```
Comment (kmrmt) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3427]: I think Boundary Value Testing needs MultiInsert maximum vectors test.
However, the test may not run because of the RAM.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 2]
```diff
@@ -3385,6 +3385,69 @@ func Test_server_MultiInsert(t *testing.T) {
...
+			- case 6.2: MultiInsert 100 vector with 50 vector with nil insert request fail
+			- case 6.3: MultiInsert 100 vector with all vector with nil insert request fail
+			- case 7.1: MultiInsert 100 vector with 1 vector with nil vector fail
+			- case 7.2: MultiInsert 100 vector with 50 vector with nil vector fail
+			- case 7.3: MultiInsert 100 vector with all vector with nil vector fail
+			- case 8.1: MultiInsert 100 vector with 1 vector with empty insert vector fail
+			- case 8.2: MultiInsert 100 vector with 50 vector with empty insert vector fail
+			- case 8.3: MultiInsert 100 vector with all vector with empty insert vector fail
+
+			*WIP*
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3441]: ?

**Review by vankichi — COMMENTED** [review]
Please check your grammar before asking review?
`success/fail` is better to set at the start of the test case for readable

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 3]
```diff
@@ -3385,6 +3385,69 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				- case 4.6: MultiInsert 100 vector with all request with empty UUID fail (vector type is float32)
+				- case 5.1: MultiInsert 100 vector with 1 vector with maximum dimension fail (vector type is uint8)
+				- case 5.2: MultiInsert 100 vector with 1 vector with maximum dimension fail (vector type is float32)
+				- case 5.3: MultiInsert 100 vector with 50 vector with maximum dimension fail (vector type is uint8)
+				- case 5.4: MultiInsert 100 vector with 50 vector with maximum dimension fail (vector type is float32)
+				- case 5.5: MultiInsert 100 vector with all vector with maximum dimension fail (vector type is uint8)
+				- case 5.6: MultiInsert 100 vector with all vector with maximum dimension fail (vector type is float32)
+
+			- float32
+				- case 5: MultiInsert 100 vector with NaN value success (vector type is float32)
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3429]: Yes it is success case (same as insert case)

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 4]
```diff
@@ -3383,6 +3383,95 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				- case 4.6: Fail to MultiInsert with all request with empty UUID (vector type is float32)
+				- case 5.1: Fail to MultiInsert with 1 vector with maximum dimension (vector type is uint8)
+				- case 5.2: Fail to MultiInsert with 1 vector with maximum dimension (vector type is float32)
+				- case 5.3: Fail to MultiInsert with 50 vector with maximum dimension (vector type is uint8)
+				- case 5.4: Fail to MultiInsert with 50 vector with maximum dimension (vector type is float32)
+				- case 5.5: Fail to MultiInsert with all vector with maximum dimension (vector type is uint8)
+				- case 5.6: Fail to MultiInsert with all vector with maximum dimension (vector type is float32)
+
+			- float32 (with 100 insert request in a single MultiInsert request)
+				- case 6: Success to MultiInsert with NaN value (vector type is float32)
```
Comment (kmrmt) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3427]: How about +-Inf, -0 case in float32?
```suggestion
				- case 6.1: Success to MultiInsert with NaN value (vector type is float32)
				- case 6.2: Success to MultiInsert with +Inf value (vector type is float32)
				- case 6.3: Success to MultiInsert with -Inf value (vector type is float32)
				- case 6.4: Success to MultiInsert with -0 value (vector type is float32)
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 5]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+
+		for j, ivec := range ivecs {
+			vec := make([]float32, dim)
+			for i := 0; i < dim; i++ {
+				vec[i] = float32(ivec[i])
+			}
+			result[j] = vec
+		}
+		return result
+	}
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3949]: could you define these functions as global in this test file?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 6]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+
 	defaultCheckFunc := func(w want, gotRes *payload.Object_Locations, err error) error {
-		if !errors.Is(err, w.err) {
-			return errors.Errorf("got_error: \"%#v\",\n\t\t\t\twant: \"%#v\"", err, w.err)
+		if !w.haveErr {
+			if !errors.Is(err, w.err) {
+				return errors.Errorf("got_error: \"%v\",\n\t\t\t\twant: \"%v\"", err, w.err)
+			}
+		} else if err == nil {
+			return errors.Errorf("got_error: nil, want non-nil error")
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4013]:
```suggestion
			return errors.Errorf("got_error: \"%v\",\n\t\t\t\twant: \"%v\"", err, w.err)
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 7]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+		return result
+	}
+
 	defaultCheckFunc := func(w want, gotRes *payload.Object_Locations, err error) error {
-		if !errors.Is(err, w.err) {
-			return errors.Errorf("got_error: \"%#v\",\n\t\t\t\twant: \"%#v\"", err, w.err)
+		if !w.haveErr {
+			if !errors.Is(err, w.err) {
+				return errors.Errorf("got_error: \"%v\",\n\t\t\t\twant: \"%v\"", err, w.err)
+			}
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4011]: IMO, I'd like to avoid using this branch.
Why you do not want to check the error.

**Review by vankichi — CHANGES_REQUESTED** [review]

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 8]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
-		       },
-		       want: want{},
-		       checkFunc: defaultCheckFunc,
-		   },
-		*/
+		{
+			name: "Equivalence Class Testing case 1.1: Success to MultiInsert 1 vector (vector type is uint8)",
+			args: args{
+				ctx:  ctx,
+				reqs: genMultiInsertReq(Uint8, vector.Gaussian, 1, intVecDim),
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4117]: I think it is ok to define 1 here as it is the test case requirement to generate 1 insert request in a multiinsert request. 
what do you think?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 9]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+			},
+			want: want{
+				wantRes: genObjectLocations(1, name, ip),
+			},
+		},
+		{
+			name: "Equivalence Class Testing case 1.3: Success to MultiInsert 100 vector (vector type is uint8)",
+			args: args{
+				ctx:  ctx,
+				reqs: genMultiInsertReq(Uint8, vector.Gaussian, 100, intVecDim),
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4151]: same as 
https://github.com/vdaas/vald/pull/1612#discussion_r876758127

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 10]
```diff
@@ -3862,19 +3862,21 @@ func Test_server_StreamInsert(t *testing.T) {
...
 		ip                string
-		ngt               service.NGT
-		eg                errgroup.Group
 		streamConcurrency int
+		svcCfg            *config.NGT
+		svcOpts           []service.Option
 	}
 	type want struct {
 		wantRes *payload.Object_Locations
+		haveErr bool // skip checking error detail
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3879]: It is because in some test cases, the error outputted is not fixed. 
The implementation of insert function will convert the requests to the map, and then insert into ngt.
https://github.com/vdaas/vald/blob/39c94dfdfc96d6559845ef6ed14cac097aa338fc/pkg/agent/core/ngt/service/ngt.go#L707
but the order of looping through the map is randomized, so I cannot predict the output error hence adding this flag to skip checking the error detail.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 7 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4011]: the reason is written here.
https://github.com/vdaas/vald/pull/1612/files#r876765568

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 11]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+						return err
+					}(),
+				},
+			}
+		}(),
+		func() test {
+			req := genMultiInsertReq(Float, vector.Gaussian, 100, f32VecDim)
+			for i := 0; i < 100; i += 2 {
+				req.Requests[i].Vector.Vector = genF32Vec(vector.Gaussian, 1, f32VecDim+1)[0]
+			}
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4434]: this is float32 test case, so I think your suggestion is not correct.
```suggestion
			for i := 0; i < len(req.Requests)/2; i++ {
				req.Requests[i].Vector.Vector = genF32Vec(vector.Gaussian, 1, f32VecDim+1)[0]
			}
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 12]
```diff
@@ -3883,85 +3885,2699 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				req.Requests[i].Vector.Id = req.Requests[0].Vector.Id
+			}
+
+			// set insert request config
+			cfg := &payload.Insert_Config{
+				SkipStrictExistCheck: false,
+			}
+			for i := 0; i < len(req.Requests); i++ {
+				req.Requests[i].Config = cfg
+			}
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:5696]: It is because `Config` can be nil, by default I think it should be nil so I explicitly set it here.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3899]: How about disassembling the error and sorting it by any new function?
Even output is not fixed, we can check it.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4117]: Yes, we should do it.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:5696]: > It is because Config can be nil, by default I think it should be nil so I explicitly set it here.

It is not a problem.
For example, it works with nil input.
```go
genMultiInsertReq := func(t objectType, dist vector.Distribution, num int, dim int, cfg ...*payload.Insert_Config) *payload.Insert_MultiRequest
....
    for i:=0; i < num; i++ {
    ...
        if cfg != nil {
            req.Requests[i].Config = cfg[0]
        }
    }
...
}
```
Or, implementing functional option works.

**Review by kmrmt — APPROVED** [review]

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: fixed here.

https://github.com/vdaas/vald/pull/1612/commits/4e32995ae28cb1597bb354f95d021482fe952df9

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: instead of defining haveErr here and sort it, I implemented a for loop to check if the output error is one of the excepted error. 

Changes:
https://github.com/vdaas/vald/pull/1612/commits/7830c38f13793444902ad193655167de6bd99760

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: fixed

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 13]
```diff
@@ -3860,106 +3860,2824 @@ func Test_server_StreamInsert(t *testing.T) {
...
+				},
+			}
+		}(),
+		func() test {
+			insertNum := 100
+			return test{
+				name: "Boundary Value Testing case 1.1: Success to MultiInsert with 0 value vector (vector type is uint8)",
+				args: args{
+					ctx:  ctx,
+					reqs: genSameVecMultiInsertReq(insertNum, []float32{0, 0, 0}, nil),
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]:
```suggestion
					reqs: genSameVecMultiInsertReq(insertNum, [intVecDim]float32, nil),
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 14]
```diff
@@ -3860,106 +3860,2824 @@ func Test_server_StreamInsert(t *testing.T) {
...
+				},
+			}
+		}(),
+		func() test {
+			insertNum := 100
+			return test{
+				name: "Boundary Value Testing case 1.2: Success to MultiInsert with 0 value vector (vector type is float32)",
+				args: args{
+					ctx:  ctx,
+					reqs: genSameVecMultiInsertReq(insertNum, []float32{0, 0, 0}, nil),
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4626]:
```suggestion
					reqs: genSameVecMultiInsertReq(insertNum, [f32VecDim]float32, nil),
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 15]
```diff
@@ -3860,106 +3860,2824 @@ func Test_server_StreamInsert(t *testing.T) {
...
+				},
+			}
+		}(),
+		func() test {
+			insertNum := 100
+			return test{
+				name: "Boundary Value Testing case 3.2: Success to MultiInsert with max value vector (vector type is float32)",
+				args: args{
+					ctx:  ctx,
+					reqs: genSameVecMultiInsertReq(insertNum, []float32{math.MaxFloat32, math.MaxFloat32, math.MaxFloat32}, nil),
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4706]:
```suggestion
					reqs: genSameVecMultiInsertReq(insertNum, [f32VecDim]float32{math.MaxFloat32, math.MaxFloat32, math.MaxFloat32}, nil),
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 13 above)
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]: It would be better to show the reason set 3 variables in a slice.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 16]
```diff
@@ -3860,106 +3860,2824 @@ func Test_server_StreamInsert(t *testing.T) {
...
+				Vector: &payload.Object_Vector{
+					Id:     "uuid-" + strconv.Itoa(i+1),
+					Vector: vec,
+				},
+				Config: cfg,
+			}
+		}
+
+		return req
+	}
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3975]: why I define it in a function is because it is the function to generate MultiInsert request which can be only used in this function. Do you think I still need to define it as a global function?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 13 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]: we cannot define like your suggestion.
I will refactor it to 
```suggestion
					reqs: genSameVecMultiInsertReq(insertNum, make([]float32, f32VecDim), nil),
```

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]: or should I create a helper function to fix it?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 17]
```diff
@@ -3860,106 +3860,2824 @@ func Test_server_StreamInsert(t *testing.T) {
...
+				},
+			}
+		}(),
+		func() test {
+			insertNum := 100
+			return test{
+				name: "Boundary Value Testing case 2.2: Success to MultiInsert with min value vector (vector type is float32)",
+				args: args{
+					ctx:  ctx,
+					reqs: genSameVecMultiInsertReq(insertNum, []float32{-math.MaxFloat32, -math.MaxFloat32, -math.MaxFloat32}, nil),
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4666]: related comment.
https://github.com/vdaas/vald/pull/1612#discussion_r879985087

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 13 above)
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]: If this function gives us any good merit, it would be better.
Either is fine..

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4606]: yes, I think we can also use the helper function for different value tests (like https://github.com/vdaas/vald/pull/1612#discussion_r879977142). 
I will fix this comment by defining the helper function for it.

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: fixed :)

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: fixed :)

**Review by vankichi — APPROVED** [review]

**Review by kmrmt — APPROVED** [review]
